aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -0,0 +1,366 @@
+
+#include <ATen/FunctionalTensorWrapper.h>
+
+#include <ATen/FunctionalInverses.h>
+#include <ATen/WrapDimUtils.h>
+#include <ATen/core/LegacyTypeDispatch.h>
+#include <c10/util/Exception.h>
+
+#include <c10/util/irange.h>
+
+namespace at {
+
+FunctionalTensorWrapper::FunctionalTensorWrapper(Tensor value)
+  : c10::TensorImpl(
+      c10::Storage(c10::make_intrusive<functionalization::FunctionalStorageImpl>(value.device(), value.numel(), value.dtype())),
+      c10::DispatchKeySet(DispatchKey::Functionalize),
+      value.dtype()
+    ),
+    value_(value),
+    // Note: "level" is a concept that we don't know how to compute in core.
+    // For now I'm retroactively setting this in functorch,
+    // but once Open Multiple Dispatch lands we should be able to calculate this in core.
+    level_(-1)
+{
+  // shallow_copy_from overwrites the storage and dispatch keyset...
+  auto functional_storage = storage_;
+  shallow_copy_from(value.getIntrusivePtr());
+  storage_ = functional_storage;
+  // TODO: should FunctionalTensorWrapper set_storage_access_should_throw()?
+  // Their storage is similar to meta tensors, which does not throw.
+  // However, users might expect to legally use storage() inside of a functionalized program, and this could be a foot gun.
+  storage_access_should_throw_ = false;
+  TORCH_INTERNAL_ASSERT(value_.defined());
+  // TODO: test using the python key to trace a function that uses the functionalization pass.
+  // I think I need this to make sure that replace_() calls hit the python key.
+  key_set_ = (value_.key_set() & c10::DispatchKeySet(c10::DispatchKeySet::FULL_AFTER, c10::DispatchKey::FuncTorchDynamicLayerBackMode))
+     | c10::DispatchKeySet(c10::DispatchKey::Functionalize);
+}
+
+void FunctionalTensorWrapper::maybe_add_update() {
+  auto storage_impl = static_cast<functionalization::FunctionalStorageImpl*>(unsafe_storage().unsafeGetStorageImpl());
+  auto updated = storage_impl->maybe_add_update(value(), view_metas_);
+  if (updated) {
+    // At this point, we know that the current tensor is up to date with its alias.
+    // Why? maybe_add_update() is only called after we've called an inplace operation on the tensor.
+    // The pre-condition in the codegen is that we always sync tensor inputs before calling a mutation op.
+    generation_ = storage_impl->generation();
+  }
+}
```

Comment:
It feels like there is a race condition here where there previously wasn't one, involving concurrent views and mutations on data. The crux of the matter is in a traditional TensorImpl, mutation on data doesn't affect tensor metadata, so you can have a situation where other threads may see the new or old floating point values, but it doesn't matter for metadata (which is stable); thus, it's safe to concurrently mutate data and take out views on the tensor. But now mutations affect tensor metadata, and now you have a data race between the metadata writes from the data mutation, and the metadata reads from the view tensor.

I suppose we could solve this by saying that there is a stricter multithreading requirement here (views and mutations are not allowed to go on concurrently) and probably in practice no one would hit it. It isn't intuitive that there would be this restriction though, and it definitely needs to be documented.